Sources/Articles/Documentation.docc/2024/beginners-guide-to-protocol-buffers-and-grpc-with-swift/beginners-guide-to-protocol-buffers-and-grpc-with-swift.md

```diff
@@ -0,0 +1,382 @@
+
+Below is the protobuf definition for the sample todo service:
+
+```protobuf
+// todo.proto
+syntax = "proto3";
+
+package todos;
+
+// 1.
+import "todo_messages.proto";
+
+// 2. 
+service TodoService {
+  // 3. 
+  rpc FetchTodos (Empty) returns (TodoList) {}
+  // 4. 
+  rpc CreateTodo (Todo) returns (Todo) {}
+  // 5. 
+  rpc DeleteTodo (TodoID) returns (Empty) {}
+  // 6. 
+  rpc CompleteTodo (TodoID) returns (Todo) {}
+}
+```
+
+1. The `todo_messages.proto` file is imported to reuse its message definitions.
+2. The `TodoService` defines the service responsible for managing todo operations.
+3. The `FetchTodos` method retrieves a list of todos, taking an empty message as input and returning a `TodoList`.
+4. The `CreateTodo` method creates a new todo, taking a `Todo` as input and returning the created `Todo`.
+5. The `DeleteTodo` method removes a todo, taking a `TodoID` as input and returning an empty response.
+6. The `CompleteTodo` method toggles the completion status of a todo, taking a `TodoID` as input and returning the updated `Todo`.
+
+It’s possible to generate the entire gRPC service protocol from this file using the protoc command. Simply run the following command:
+
+```sh
+protoc --swift_out=./ --grpc-swift_out=./ todo.proto
+```
+
+This will generate the `todo.grpc.swift` file, which contains the protocols that must be implemented when building the gRPC server. It's also worth mentioning that the `protoc` command generates both client and server-side protocols by default.
+
+The process that we just described is an ahead-of-time (manual) code generation, quite similar what we have for the [Swift OpenAPI generator](https://github.com/apple/swift-openapi-generator/blob/main/Examples/README.md#ahead-of-time-manual-code-generation) tool. 
+
+The generated files depend on two Swift libraries, but these packages also include Swift Package Plugins. By leveraging these plugins, the entire generation process can be integrated into the build pipeline.
+
+Now, let's explore how to use the [Swift Protobuf](https://github.com/apple/swift-protobuf) and the [gRPC Swift](https://github.com/grpc/grpc-swift) libraries to set up a basic server.
+
+## Using Protocol Buffers and gRPC with Swift
+
+First, create a new Swift Package and add both the Protobuf and gRPC libraries as package dependencies. The [Swift Argument Parser](https://github.com/apple/swift-argument-parser) will be included too, and the Protobuf & gRPC plugins will be connected to the executable target. Here’s how to set it up in your `Package.swift` file:
+
+
+```swift
+// swift-tools-version:5.10
+import PackageDescription
+
+let package = Package(
+    name: "grpc-example",
+    platforms: [
+        .macOS(.v14),
+    ],
+    dependencies: [
+        .package(url: "https://github.com/apple/swift-argument-parser.git", from: "1.4.0"),
+        .package(url: "https://github.com/apple/swift-protobuf.git", from: "1.28.0"),
```

Comment:
Can we change this article to swift-protobuf v2?

Reply:
Done.